packages/astro/src/core/middleware/callMiddleware.ts

```diff
@@ -38,13 +44,28 @@ import { AstroError, AstroErrorData } from '../errors/index.js';
 export async function callMiddleware(
 	onRequest: MiddlewareHandler,
 	apiContext: APIContext,
-	responseFunction: () => Promise<Response> | Response
+	responseFunction: (
+		apiContext: APIContext,
+		reroutePayload?: RewritePayload
+	) => Promise<Response> | Response,
+	// TODO: remove these two arguments once rerouting goes out of experimental
+	enableRerouting: boolean,
+	logger: Logger
 ): Promise<Response> {
 	let nextCalled = false;
 	let responseFunctionPromise: Promise<Response> | Response | undefined = undefined;
-	const next: MiddlewareNext = async () => {
+	const next: MiddlewareNext = async (payload) => {
 		nextCalled = true;
-		responseFunctionPromise = responseFunction();
+		if (enableRerouting) {
+			responseFunctionPromise = responseFunction(apiContext, payload);
+		} else {
+			logger.warn(
+				'router',
+				'You tried to use the routing feature without enabling it via experimental flag. This is not allowed.'
```

Comment:
```suggestion
				'The rewrite API is experimental. To use this feature, add the `rewriting` flag to the `experimental` object in your Astro config.'
```
Flipped the phrasing to guide the user to the right configuration.